drivers/sensor/qdec_mcux/qdec_mcux.c

```diff
@@ -0,0 +1,204 @@
+
+struct qdec_mcux_config {
+	ENC_Type *base;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif /* CONFIG_PINCTRL */
+	size_t xbar_maps_len;
+	int xbar_maps[];
+};
+
+struct qdec_mcux_data {
+	enc_config_t qdec_config;
+	int32_t position;
+	int16_t difference;
+	int16_t revolution;
+};
+
+static int qdec_mcux_attr_set(const struct device *dev, enum sensor_channel ch,
+	enum sensor_attribute attr, const struct sensor_value *val)
+{
+	const struct qdec_mcux_config *config = dev->config;
+	struct qdec_mcux_data *data = dev->data;
+
+	if (ch != SENSOR_CHAN_ROTATION) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_QDEC_MOD_VAL:
+		data->qdec_config.positionModulusValue = val->val1;
+		ENC_Init(config->base, &data->qdec_config);
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int qdec_mcux_attr_get(const struct device *dev, enum sensor_channel ch,
+	enum sensor_attribute attr, struct sensor_value *val)
+{
+	struct qdec_mcux_data *data = dev->data;
+
+	if (ch != SENSOR_CHAN_ROTATION) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_QDEC_MOD_VAL:
+		/* NOTE: Register is an unsigned 32 bit value which is stored
+		 * in a signed 32 bit integer
+		 */
+		val->val1 = data->qdec_config.positionModulusValue;
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int qdec_mcux_fetch(const struct device *dev, enum sensor_channel ch)
+{
+	const struct qdec_mcux_config *config = dev->config;
+	struct qdec_mcux_data *data = dev->data;
+
+	if (ch != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	/* Read position */
+	data->position = ENC_GetPositionValue(config->base);
+	/* Read hold values to get the values from when position was read */
+	data->difference = ENC_GetHoldPositionDifferenceValue(config->base);
+	data->revolution = ENC_GetHoldRevolutionValue(config->base);
+
+	LOG_DBG("pos %d, dif %d, rev %d",
+		data->position, data->difference, data->revolution);
+
+	return 0;
+}
+
+static int qdec_mcux_ch_get(const struct device *dev, enum sensor_channel ch,
+			 struct sensor_value *val)
+{
+	struct qdec_mcux_data *data = dev->data;
+
+	switch (ch) {
+	case SENSOR_CHAN_ROTATION:
+		val->val1 = (int64_t)(data->position * 360) /
+			    data->qdec_config.positionModulusValue;
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_RPM:
+		val->val1 = data->revolution;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api qdec_mcux_api = {
+	.attr_set = &qdec_mcux_attr_set,
+	.attr_get = &qdec_mcux_attr_get,
+	.sample_fetch = &qdec_mcux_fetch,
+	.channel_get = &qdec_mcux_ch_get,
+};
+
+static void init_inputs(const struct device *dev)
+{
+	int i;
+	const struct qdec_mcux_config *config = dev->config;
+
+#ifdef CONFIG_PINCTRL
+	i = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	assert(i == 0);
+#endif
+
+	/* Quadrature Encoder inputs are only accessible via crossbar */
+	XBARA_Init(XBARA1);
```

Comment:
Can `XBARA1` be read from Devicetree as this can change to `XBARA` on certain platforms.

Reply:
I added it to the overlay. I would like to derive it e.g. from a xbar function but was unable to.

Reply:
@jeppenodgaard the `xbar` property has been added as part of `nxp,mcux-qdec.yaml` binding. Can this separated out to a separate node with its own binding inside `dts/bindings/arm`?

Reply:
I think it would be nice to have a xbar driver. Then maybe inside the qdec node there would be a xbar child node that would be setup by the xbar driver so the qdec driver is more clean. Maybe this could be a seperate issue/PR?

Anyway if you want me to add it to this PR:
Do you want me to create e.g. `dts/bindings/arm/nxp,mcux-xbar.yaml` with a xbar property? I should also probably move xbar-maps?

Reply:
I agree that having a driver would be a separate PR. I was thinking if we could add the device tree bindings as you have proposed above.

Reply:
@mmahadevan108 I have tried to add your suggestion.